egi_notebooks_hub/d4science.py

```diff
@@ -151,3 +152,193 @@ def pre_spawn_start(self, user, spawner):
+            key=key,
+            audience=self.client_id,
+            algorithms=["RS256"],
+        )
+        # TODO: add extra checks?
+        permissions = decoded_token["authorization"]["permissions"]
+        user_data["auth_state"].update(
+            {
+                "uma_token": token,
+                "permissions": permissions,
+                "context": context,
+            }
+        )
+        return user_data
+
+    async def pre_spawn_start(self, user, spawner):
+        """Pass gcube-token to spawner via environment variable"""
+        auth_state = await user.get_auth_state()
+        if not auth_state:
+            # auth_state not enabled
+            return
+        spawner.environment["GCUBE_TOKEN"] = auth_state["uma_token"]
+        # spawner.environment["DATAMINER_URL"] = auth_state["wps-endpoint"]
+        spawner.environment["GCUBE_VRE"] = unquote(auth_state["context"])
+
+
+class D4ScienceSpawner(KubeSpawner):
+    frame_ancestors = Unicode(
+        "https://*.d4science.org 'self'",
+        config=True,
+        help="""Frame ancestors for embedding the hub in d4science""",
+    )
+    sidecar_image = Unicode(
+        "eginotebooks/d4science-storage",
+        config=True,
+        help="""the D4science storage image to use""",
+    )
+
+    def get_args(self):
+        args = super().get_args()
+        tornado_settings = {
+            "headers": {
+                "Content-Security-Policy": "frame-ancestors %s" % self.frame_ancestors
+            },
+            "cookie_options": {"samesite": "None", "secure": True},
+        }
+        # TODO: check if this keeps making sense
+        return [
+            "--SingleUserNotebookApp.tornado_settings=%s" % tornado_settings,
+            "--FileCheckpoints.checkpoint_dir='/home/jovyan/.notebookCheckpoints'",
+            "--FileContentsManager.use_atomic_writing=False",
+            "--NotebookApp.ResourceUseDisplay.track_cpu_percent=True",
+        ] + args
+
+    def auth_state_hook(self, spawner, auth_state):
+        permissions = auth_state["permissions"]
+        # this will filter according to permissions
+        # Assuming that there are no permissions coming from D4Science,
+        # the everything is allowed
+        if spawner.profile_list and permissions:
+            allowed_profiles = [claim["rsname"] for claim in permissions]
+            self.log.debug("allowed profiles: %s", allowed_profiles)
+            spawner.profile_list = list(
+                filter(
+                    lambda x: x.get("profile_type", None) in allowed_profiles,
+                    self.profile_list,
+                )
+            )
+
+    async def pre_spawn_hook(self, spawner):
+        gcube_token = spawner.environment.get("GCUBE_TOKEN", "")
+        vre = spawner.environment.get("GCUBE_VRE", "")
+        if vre:
+            vre = vre[vre.rindex("/") + 1 :]
+            spawner.log.info("VRE: %s", vre)
+            spawner.environment["VRE"] = vre
```

Comment:
as we would like to add the VRE as pod annotation, can this be done here?

Reply:
the spawner has this https://github.com/jupyterhub/kubespawner/blob/master/kubespawner/spawner.py#L651, so could be added here yes

Reply:
ok i will add the impl here